infra/experimental/SystemSan/SystemSan.cpp

```diff
@@ -324,44 +330,56 @@ void inspect_for_arbitrary_file_open(pid_t pid, const user_regs_struct &regs) {
 }
 
 std::string read_evil_link_bombfile() {
-    const std::ifstream bombfile(kEvilLinkBombfile, std::ios_base::binary);
+    const std::ifstream bombfile(kEvilLinkBombfile,
+                                 std::ios_base::binary);
     if (bombfile.fail())
       return "";
     std::stringstream stream;
     stream << bombfile.rdbuf();
     return stream.str();
 }
 
+// https://oss-fuzz.com/testcase-detail/4882113260552192
+void report_bug_in_process(std::string bug_type, pid_t pid) {
+  std::cerr << "===BUG DETECTED: " << bug_type << "===" << std::endl;
+  tgkill(root_pids[pid].parent_tid, pid, SIGABRT);
+}
+
 void inspect_for_evil_link(pid_t pid, const user_regs_struct &regs) {
   (void) regs;
   std::string contents = read_evil_link_bombfile();
-  if ((contents.compare("original")) != 0) {
-    report_bug(kEvilLinkError, pid);
-    exit(1);
+  if ((contents.compare(kEvilLinkBombfileContents)) != 0) {
+
+    report_bug_in_process(kEvilLinkError, pid);
   }
 }
 
-void evil_open_hook(pid_t pid, const user_regs_struct &regs) {
-  // Inspect a PID's register for the sign of arbitrary file open.
-  std::string path = read_string(pid, regs.rsi, kRootDirMaxLength);
+void evil_openat_hook(pid_t pid, const user_regs_struct &regs) {
+  std::string path = read_string(pid, regs.rsi, kPathMax);
   if (!path.length()) {
     return;
   }
+  if (std::filesystem::exists(path))
+    return;
   size_t slash_idx = path.rfind('/');
   if (slash_idx == std::string::npos)
     return;
 
   std::string dir = path.substr(0, slash_idx);
-  std::cout << "DIR" << slash_idx << " " << dir << std::endl;
   if ((dir.compare("/tmp")) != 0)
     return;
-  std::string command = "ln -s " + kEvilLinkBombfile + " " + path;
-  std::cout << "COMMAND" << command << std::endl;
+
+  std::string command = "rm -f " + path + " && ln -s " + kEvilLinkBombfile + " " + path;
+  std::cout << "COMMAND " << command << std::endl;
   system(command.c_str());
 }
 
 void initialize_evil_link_bombfile() {
-  system(("printf 'original' > " + kEvilLinkBombfile).c_str());
+  std::string command = ("printf " + kEvilLinkBombfileContents + " > " +
```

Comment:
Can we add a TODO to write this in C++ rather then shelling out? 

Same for above in evil_openat_hook.

Reply:
haha sure.